notebooks/nearest_neighbors.livemd

```diff
@@ -0,0 +1,255 @@
+
+Here's another, more abstract example.
+
+| ![KDtree](https://upload.wikimedia.org/wikipedia/commons/4/4f/Kd-tree-example-with_records_as_leaf_nodes.png) |
+| :-----------------------------------------------------------------------------------------------------------: |
+
+Figure 2: A different example of 3-dimensional KDtree.
+
+<!-- livebook:{"break_markdown":true} -->
+
+In this example, the dimensions are abstract values like name, salary, and age. However, because we can define a linear order for each feature (such as a natural order for salary and age, and a lexicographic order for names), we can treat them like any other metric space.
+
+Now, let's move on to the code and implementation of KDTrees in Scholar!
+
+<!-- livebook:{"branch_parent_index":3} -->
+
+## KDtrees in action - Practical Example
+
+In this example, we'll use data from the [National Health and Nutrition Examination Survey (NHANES) 2013-2014](https://archive.ics.uci.edu/dataset/887/national+health+and+nutrition+health+survey+2013-2014+(nhanes)+age+prediction+subset). The dataset consists of 7 features that are biomedical indicators. For more information, click the link above.
+
+Now, let's read the data into an Explorer dataframe.
+
+```elixir
+digits_url =
+  "https://archive.ics.uci.edu/static/public/887/national+health+and+nutrition+health+survey+2013-2014+(nhanes)+age+prediction+subset.zip"
+
+data =
+  Req.get!(digits_url).body
+
+{:ok, [{_, data}]} = :zip.extract(data, [:memory])
+df_data = DF.load_csv!(data)
+```
+
+In this analysis, we want to determine if the most similar people, in terms of medical parameters, belong to the same age group. In the dataset, we categorize people into two groups: adults (18-64 years) and seniors (65+ years). Intuitively, we might expect that people with similar medical results are often in the same age group, but we want to quantify the extent to which this is true. To do this, we compute the 5 closest results for each patient and then calculate the average for these two groups.
+
+First, we need to separate the medical results from the data that indicates the age group.
+
+```elixir
+x = df_data |> DF.discard(["age_group", "SEQN", "RIDAGEYR"]) |> Nx.stack(axis: 1)
+y = Nx.stack(S.cast(df_data["age_group"], :category), axis: 1)
+
+x = Scholar.Preprocessing.StandardScaler.fit_transform(x)
+
+{x, y}
+```
+
+Once the data is preprocessed, we can move on to the KDTree model. First, we need to set up the model using `Scholar.Neighbors.KDTree.fit`. This method initializes the tree structure for further processing.
+
+```elixir
+num_neighbors = 6
+
+knn_model =
+  EXLA.jit_apply(&Scholar.Neighbors.KDTree.fit(&1, k: num_neighbors, metric: :euclidean), [x])
```

Comment:
I tend to prefer jit, a bit less confusing. :)

```suggestion
  EXLA.jit(&Scholar.Neighbors.KDTree.fit(&1, k: num_neighbors, metric: :euclidean).(x)
```

Can we change it accordingly?

Reply:
Ok, sure